pkg/ccl/changefeedccl/sink_pubsub.go

```diff
@@ -0,0 +1,526 @@
+
+	// flushDone channel signaled when flushing completes.
+	flushDone chan struct{}
+
+	// errChan is written to indicate an error while sending message.
+	errChan chan error
+
+	withTopicName string
+
+	client pubsubClient
+}
+
+// my idea is to force this to be an interface so it adheres to api, but not too sure since Sink is already an API
+//func getPubsubSinkObject(url sinkURL, withTopicName bool) pubsubSink {
+//	return &pubsubSinkObject{url: url, withTopicName: withTopicName}
+//}
+
+// getGCPCredentials returns gcp credentials parsed out from url
+func getGCPCredentials(ctx context.Context, u sinkURL) (*google.Credentials, error) {
+	const authParam = "AUTH"
+	const authSpecified = "specified"
+	const authImplicit = "implicit"
+	const authDefault = "default"
+
+	var credsJSON []byte
+	var creds *google.Credentials
+	var err error
+	authOption := u.consumeParam(authParam)
+
+	// implemented according to https://github.com/cockroachdb/cockroach/pull/64737
+	switch authOption {
+	case authImplicit:
+		creds, err = google.FindDefaultCredentials(ctx, gcpScope)
+		if err != nil {
+			return nil, err
+		}
+		return creds, nil
+	case authSpecified:
+		fallthrough
+	case authDefault:
+		fallthrough
+	default:
+		err := u.decodeBase64(credentialsParam, &credsJSON)
+		if err != nil {
+			return nil, errors.Wrap(err, "decoding credentials json")
+		}
+		creds, err = google.CredentialsFromJSON(ctx, credsJSON, gcpScope)
+		if err != nil {
+			return nil, errors.Wrap(err, "creating credentials")
+		}
+		return creds, nil
+	}
+}
+
+// MakePubsubSink returns the corresponding pubsub sink based on the url given
+func MakePubsubSink(
+	ctx context.Context, u *url.URL, opts map[string]string, targets jobspb.ChangefeedTargets,
+) (Sink, error) {
+
+	pubsubURL := sinkURL{URL: u, q: u.Query()}
+	pubsubTopicName := pubsubURL.consumeParam(changefeedbase.SinkParamTopicName)
+
+	switch changefeedbase.FormatType(opts[changefeedbase.OptFormat]) {
+	case changefeedbase.OptFormatJSON:
+	default:
+		return nil, errors.Errorf(`this sink is incompatible with %s=%s`,
+			changefeedbase.OptFormat, opts[changefeedbase.OptFormat])
+	}
+
+	switch changefeedbase.EnvelopeType(opts[changefeedbase.OptEnvelope]) {
+	case changefeedbase.OptEnvelopeWrapped:
+	default:
+		return nil, errors.Errorf(`this sink is incompatible with %s=%s`,
+			changefeedbase.OptEnvelope, opts[changefeedbase.OptEnvelope])
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	p := &pubsubSink{
+		workerCtx: ctx, numWorkers: numOfWorkers,
+		exitWorkers: cancel, withTopicName: pubsubTopicName,
+	}
+
+	// creates custom pubsub object based on scheme
+	switch u.Scheme {
+	case gcpScheme:
+		const regionParam = "region"
+		projectID := pubsubURL.Host
+		region := pubsubURL.consumeParam(regionParam)
+		g := &gcpPubsubClient{topics: p.getTopicsMap(targets, pubsubTopicName),
+			ctx: ctx, projectID: projectID, region: region, url: pubsubURL}
+		p.client = g
+		return p, nil
+	default:
+		_ = p.Close()
```

Comment:
Why do we have to do it? Is it to avoid leaking cancel?  Perhaps we can do that in the Dial instead?